harness/tests/launch/test_autodeepspeed.py

```diff
@@ -0,0 +1,359 @@
+import contextlib
+import os
+import time
+import unittest.mock as mock
+from typing import Any, Iterator, List
+
+import pytest
+
+from determined import ClusterInfo, RendezvousInfo, constants
+from determined.launch import autodeepspeed
+
+
+@mock.patch("subprocess.Popen")
+@mock.patch("determined.get_cluster_info")
+@mock.patch("determined.util.check_sshd")
+@mock.patch("time.time")
+def test_launch_multi_slot_chief(
+    mock_time: mock.MagicMock,
+    mock_check_sshd: mock.MagicMock,
+    mock_cluster_info: mock.MagicMock,
+    mock_subprocess: mock.MagicMock,
+) -> None:
+    cluster_info = make_mock_cluster_info(["0.0.0.0", "0.0.0.1"], 0)
+    mock_cluster_info.return_value = cluster_info
+    mock_start_time = time.time()
+    mock_time.return_value = mock_start_time
+    train_entrypoint = "model_def:TrialClass"
+    sshd_cmd = [
+        "/usr/sbin/sshd",
+        "-p",
+        str(constants.DTRAIN_SSH_PORT),
+        "-f",
+        "/run/determined/ssh/sshd_config",
+        "-D",
+    ]
+
+    pid_server_cmd = [
```

Comment:
Also, these launch commands look like an alarming amount of code duplication, maybe break some of these up into helper functions in `autodeepspeed.py`, and call those from here or something?